### PR TITLE
Update ref based rna-seq

### PR DESCRIPTION
**Pull request by lldelisle**
This PR in based on #3272 
Additional changes in this PR:
- Use collection instead of individual datasets
- Use last up to date version of tools (except STAR which I could do).
- Add myself as contributor
- Update some phrasing: changing 'absolute fold-change > 2' to 'fold-change > 2 or < 0.5' etc.
- Change color scale of log2(counts)
- Change some images / text on strandness
- Add a step to check Pasilla gene status
- Propose an alternative with counts with STAR
- Use a choose your own adventure (CYOA) style thanks to @hexylena
- Use last version of heatmap2 and plot z-score without computing them. Moved the part on how to compute z-scores in a detailed section and do it on all genes. 
- Use multiQC on cutadapt thanks to @bebatut 
Aside:
- I updated create_new file to set name at creation.
- @hexylena added a post on CYOA and updated 'Create new tutorial'.
- @bebatut updated copy datasets between histories

### CODE REVIEW

**Comment by lldelisle** [comment]
TODO:
- Keep group tag but say it is an alternative to collections.
- Check Pasilla gene with ID just after DESeq2

TODO if time:
- update heatmap2 tool to allow Z-score computation before clustering.

**Comment by lldelisle** [comment]
Ready for review @hrhotz 
@bebatut would you mind to check that the changes on strandness are ok for you (if you have time I am also interested in your review for the rest but this is the biggest change I made on what you wrote)?

File: topics/transcriptomics/tutorials/ref-based/tutorial.md [hunk 1]
```diff
@@ -615,17 +624,19 @@ To compare the expression of single genes between different conditions (*e.g.* w
 > {: .solution}
 {: .question}
 
-Two main tools are available for read counting: [**HTSeq-count**](http://htseq.readthedocs.io/en/release_0.9.1/count.html) ({% cite anders2015htseq %}) or **featureCounts** ({% cite liao2013featurecounts %}). **featureCounts** is considerably faster and requires far less computational resources, so we will use it here.
+Two main tools are available for read counting: [**HTSeq-count**](http://htseq.readthedocs.io/en/release_0.9.1/count.html) ({% cite anders2015htseq %}) or **featureCounts** ({% cite liao2013featurecounts %}). **STAR** allows to count reads while mapping. Its results are identical to **HTSeq-count**. While this output is sufficient for most analyses, **featureCounts** offers more customization on how to count reads (minimum mapping quality, counting reads instead of fragments, count transcripts instead of genes etc.).
```
Comment (nsoranzo) [inline topics/transcriptomics/tutorials/ref-based/tutorial.md:627]:
```suggestion
Two main tools are available for read counting: [**HTSeq-count**](http://htseq.readthedocs.io/en/release_0.9.1/count.html) ({% cite anders2015htseq %}) or **featureCounts** ({% cite liao2013featurecounts %}). Additionally, **STAR** allows to count reads while mapping: its results are identical to those from **HTSeq-count**. While this output is sufficient for most analyses, **featureCounts** offers more customization on how to count reads (minimum mapping quality, counting reads instead of fragments, count transcripts instead of genes etc.).
```

File: topics/transcriptomics/tutorials/ref-based/tutorial.md [hunk 2]
```diff
@@ -645,56 +656,145 @@ Some library preparation protocols create so-called *stranded* RNA-Seq libraries
...
+There are 3 ways to estimate strandness from **STAR** results:
+First, we can do a visual inspection of read strands on IGV (for Paired-end dataset it is less easy than with single read and when you have a lot of samples, this can be painful). Second, you can use the output of **STAR** with the counts. Another option is to estimate these parameters with a tool called **Infer Experiment** from the RSeQC ({% cite wang2012rseqc %}) tool suite. This tool takes the BAM files from the mapping, selects a subsample of the reads and compares their genome coordinates and strands with those of the reference gene model (from an annotation file). Based on the strand of the genes, it can gauge whether sequencing is strand-specific, and if so, how reads are stranded (forward or reverse).
 
-> ### {% icon hands_on %} Hands-on: Determining the library strandness
+You don't need to do the three.
```
Comment (nsoranzo) [inline topics/transcriptomics/tutorials/ref-based/tutorial.md:664]: You can choose between the three following hands-on.

File: topics/transcriptomics/tutorials/ref-based/tutorial.md [hunk 3]
```diff
@@ -645,56 +656,145 @@ Some library preparation protocols create so-called *stranded* RNA-Seq libraries
...
-> >
-> > 1. Results for `GSM46177`:
-> >
-> >    ```
-> >    This is PairEnd Data
-> >    Fraction of reads failed to determine: 0.0963
-> >    Fraction of reads explained by "1++,1--,2+-,2-+": 0.4649
-> >    Fraction of reads explained by "1+-,1-+,2++,2--": 0.4388
-> >    ```
+>    > ### {% icon comment %} Interprete the output of Infer Experiment
```
Comment (nsoranzo) [inline topics/transcriptomics/tutorials/ref-based/tutorial.md:757]:
```suggestion
>    > ### {% icon comment %} Interpret the output of Infer Experiment
```

**Comment by lldelisle** [comment]
Thanks @nsoranzo

File: topics/transcriptomics/tutorials/ref-based/tutorial.md [hunk 4]
```diff
@@ -756,28 +881,87 @@ We now run **featureCounts** to count the number of reads per annotated gene.
...
+{: .hands_on.STAR}
+
+> ### {% icon question %} Questions
+>
+> 1. How many reads are unmapped/multi-mapped?
+> 2. At which line starts gene counts?
+>
+> > ### {% icon solution %} Solution
+> >
+> > 1. There are 1'190'029 unmapped reads and 571'324 unmapped reads.
```
[the remark this answers is not in the document]
Comment (lldelisle) [reply topics/transcriptomics/tutorials/ref-based/tutorial.md]: Thanks

**Comment by lldelisle** [comment]
I am still working on this PR (just the heatmap part because I just updated heatmap2 tool so it can compute the z-score).

**Comment by lldelisle** [comment]
Heatmap2 changes done.
Waiting for last reviews before changing the workflow.

File: topics/transcriptomics/tutorials/ref-based/tutorial.md [hunk 5]
```diff
@@ -756,6 +889,76 @@ We now run **featureCounts** to count the number of reads per annotated gene.
...
+Later on the tutorial we will need to get the size of each gene. This is one of the output of **FeatureCounts** but we can also obtain it directly from the gene annotation file. As this is quite long, we recommand to launch it now.
+
+> ### {% icon hands_on %} Hands-on: Getting gene length
+>
+> 1. {% tool [Gene length and GC content](toolshed.g2.bx.psu.edu/repos/iuc/length_and_gc_content/length_and_gc_content/0.1.2) %} with the following parameters:
+>    - *"Select a built-in GTF file or one from your history"*: `Use a GTF from history`
+>      - {% icon param-file %} *"Select a GTF file"*: `Drosophila_melanogaster.BDGP6.87.gtf`
+>    - *"Analysis to perform"*: `gene lengths only`
+>
+>    > > ### {% icon warning %} Check the version of the tool below.
```
Comment (gallardoalba) [inline topics/transcriptomics/tutorials/ref-based/tutorial.md:953]:
```suggestion
>    > ### {% icon warning %} Check the version of the tool below.
```

Comment (gallardoalba) [reply topics/transcriptomics/tutorials/ref-based/tutorial.md:953]: It doesn't render correctly; I think that it fixes the problem.

Screenshot from 2022-04-14 12-30-55

File: topics/transcriptomics/tutorials/ref-based/tutorial.md [hunk 6]
```diff
@@ -1129,37 +1335,79 @@ Here, treatment is the primary factor that we are interested in. The sequencing
...
+>    - *"Choice of Input data"*: `Count data (e.g. from HTSeq-count, featureCounts or StringTie)`
+>    - In *"Output options"*:
+>        - *"Output selector"*: `Generate plots for visualizing the analysis results`, `Output normalised counts`
 >
 >    > ### {% icon comment %} Comment: Using group tags for large sample sets
 >    >
->    > If you have a large number of samples, or a complex experimental design, manually selecting files for each factor level
->    > may be a lot of work. In these situations, **group tags** can make your life a lot easier.
+>    > If you have a large number of samples, or a complex experimental design, an alternative to collections is **group tags**.
 >    >
```
Comment (gallardoalba) [inline topics/transcriptomics/tutorials/ref-based/tutorial.md:1411]: The new image seems to be too small; I think that it would be nice to increase the resolution.

Screenshot from 2022-04-14 12-41-08

Comment (gallardoalba) [reply topics/transcriptomics/tutorials/ref-based/tutorial.md:1411]: Same for this one:

Screenshot from 2022-04-14 12-42-12

File: topics/transcriptomics/tutorials/ref-based/tutorial.md [hunk 7]
```diff
@@ -1561,43 +1813,65 @@ We have extracted genes that are differentially expressed in treated (PS gene-de
...
+>   - {% icon param-collection %} *"Input List"*: `featureCounts on collection N: Feature lengths`
+>   - *"How should a dataset be selected?"*: `The first dataset`
+> </div>
 >
-> 6. {% tool [Change Case](ChangeCase) %} with
->    - {% icon param-file %} *"From"*: the `feature lengths` (output of **featureCounts**)
+> <div class="STAR" markdown="1">
+> 1. Copy the output of **Gene length and GC content** {% icon tool %} (`Gene length`) into this history
+>
+>   {% snippet faqs/galaxy/histories_copy_dataset.md %}
```
Comment (gallardoalba) [inline topics/transcriptomics/tutorials/ref-based/tutorial.md:1853]: It seems that this snippet doesn't render correctly.

Screenshot from 2022-04-14 12-50-09

File: topics/transcriptomics/tutorials/ref-based/tutorial.md (hunk 5 above)
Comment (lldelisle) [reply topics/transcriptomics/tutorials/ref-based/tutorial.md]: Thanks I had found it but forgot to push.

File: topics/transcriptomics/tutorials/ref-based/tutorial.md (hunk 7 above)
Comment (lldelisle) [reply topics/transcriptomics/tutorials/ref-based/tutorial.md:1853]: Good catch

Comment (lldelisle) [reply topics/transcriptomics/tutorials/ref-based/tutorial.md:1853]: A space was missing

File: topics/transcriptomics/tutorials/ref-based/tutorial.md (hunk 6 above)
Comment (lldelisle) [reply topics/transcriptomics/tutorials/ref-based/tutorial.md:1411]: This happened when I converted pdf to png. I will redo the conversion.

File: topics/transcriptomics/tutorials/ref-based/tutorial.md [hunk 8]
```diff
@@ -116,17 +117,13 @@ In the second part of the tutorial, read counts of all 7 samples are used to ide
...
->
->    {% snippet faqs/galaxy/datasets_rename.md %}
->
-> 4. Check that the datatype is `fastqsanger` (e.g. **not** `fastq`). If it is not, please change the datatype to `fastqsanger`.
+> 3. Check that the datatype is `fastqsanger` (e.g. **not** `fastq`). If it is not, please change the datatype to `fastqsanger`.
 >
 >    {% snippet faqs/galaxy/datasets_change_datatype.md datatype="fastqsanger" %}
 >
-> 5. Add to each dataset a tag corresponding to the name of the sample (`#GSM461177` or `#GSM461180`)
+> 4. Create a paired collection named `2 PE fastqs`, rename your pairs with the sample name followed by the attributes: `GSM461177_untreat_paired` and `GSM461180_treat_paired`.
```
Comment (gallardoalba) [inline topics/transcriptomics/tutorials/ref-based/tutorial.md:124]: Does make sense to simplify the little bit sample names? I'm not sure if including "_paired" provides meaningful information at all.

File: topics/contributing/tutorials/create-new-tutorial-content/tutorial.md [hunk 9]
```diff
@@ -965,6 +965,63 @@ And in your text you can use braces to refer to the term
 > {: .code-out}
 {: .code-2col}
 
+## Choose Your Own Tutorial
```
Comment (gallardoalba) [inline topics/contributing/tutorials/create-new-tutorial-content/tutorial.md:968]: I think that it should be included in a different PR, since it doesn't correspond to the Ref-based RNA-seq tutorial.

File: news/_posts/2022-04-12-cyot.md [hunk 10]
```diff
@@ -0,0 +1,14 @@
+---
+title: "New Tutorial Feature: Choose Your Own Tutorial"
```
Comment (gallardoalba) [inline news/_posts/2022-04-12-cyot.md:2]: Also probably it would be better to include it in a different PR.

File: topics/transcriptomics/tutorials/ref-based/tutorial.md [hunk 11]
```diff
@@ -1735,15 +2009,13 @@ Here we would like to visualize the 2 KEGG pathways: the over-represented `00010
...
+> 3. Create a new tabular file from the following (IDs of the pathways to plot) named `KEGG pathways to plot`
 >
 >    ```
 >    00010
 >    03040
 >    ```
 >
-> 4. Rename to `KEGG pathways to plot`
->
 > 5. {% tool [Pathview](toolshed.g2.bx.psu.edu/repos/iuc/pathview/pathview/1.24.0+galaxy0) %} with
```
Comment (gallardoalba) [inline topics/transcriptomics/tutorials/ref-based/tutorial.md:2019]:
```suggestion
> 4. {% tool [Pathview](toolshed.g2.bx.psu.edu/repos/iuc/pathview/pathview/1.24.0+galaxy0) %} with
```

File: topics/transcriptomics/tutorials/ref-based/tutorial.md [hunk 12]
```diff
@@ -1561,43 +1813,65 @@ We have extracted genes that are differentially expressed in treated (PS gene-de
...
+>
+> 1. {% tool [Change Case](ChangeCase) %} with the following parameters:
+>
+>    - {% icon param-file %} *"From"*: <span class="featureCounts" markdown="1">`GSM461177_untreat_paired` (output of **Extract Dataset** {% icon tool %})</span><span class="STAR" markdown="1">`Gene length`</span>
 >    - *"Change case of columns"*: `c1`
 >    - *"Delimited by"*: `Tab`
 >    - *"To"*: `Upper case`
 >
-> 7. Rename the output to `Gene IDs and length`
+> 8. Rename the output to `Gene IDs and length`
```
Comment (gallardoalba) [inline topics/transcriptomics/tutorials/ref-based/tutorial.md:1867]:
```suggestion
> 2. Rename the output to `Gene IDs and length`
```

File: topics/transcriptomics/tutorials/ref-based/tutorial.md [hunk 13]
```diff
@@ -1561,43 +1813,65 @@ We have extracted genes that are differentially expressed in treated (PS gene-de
...
+>   {% snippet faqs/galaxy/histories_copy_dataset.md %}
+> </div>
+>
+{: .hands_on}
+
+> ### {% icon hands_on %} Hands-on: Adapt the gene length to goseq
+>
+> 1. {% tool [Change Case](ChangeCase) %} with the following parameters:
+>
+>    - {% icon param-file %} *"From"*: <span class="featureCounts" markdown="1">`GSM461177_untreat_paired` (output of **Extract Dataset** {% icon tool %})</span><span class="STAR" markdown="1">`Gene length`</span>
```
Comment (gallardoalba) [inline topics/transcriptomics/tutorials/ref-based/tutorial.md:1862]: Probably there is a good reason for that, but I cannot understand why you are using HTML code here.

Comment (hexylena) [reply topics/transcriptomics/tutorials/ref-based/tutorial.md:1862]: it's for the choose your own tutorial feature.

File: topics/transcriptomics/tutorials/ref-based/tutorial.md (hunk 8 above)
Comment (gallardoalba) [reply topics/transcriptomics/tutorials/ref-based/tutorial.md:124]: Ahh I see, you are also using single datasets; then it is completely fine.

Comment (gallardoalba) [reply topics/transcriptomics/tutorials/ref-based/tutorial.md:124]: Perhaps using labels and simplifying the sample names could make the tutorial easier to follow.

File: topics/transcriptomics/tutorials/ref-based/tutorial.md [hunk 14]
```diff
@@ -1129,37 +1335,79 @@ Here, treatment is the primary factor that we are interested in. The sequencing
...
+> 3. {% tool [Search in textfiles (grep)](toolshed.g2.bx.psu.edu/repos/bgruening/text_processing/tp_grep_tool/1.1.1) %} with the following parameters:
+>    - {% icon param-file %} *"Select lines from"*: output of **Extract element identifiers** {% icon tool %}
+>    - *"Regular Expression"*: `single`
+>
+> 4. {% tool [Filter collection](__FILTER_FROM_FILE__) %} with the following parameters:
+>    - {% icon param-collection %} *"Input Collection"*: `all counts`
+>    - *"How should the elements to remove be determined?"*: `Remove if identifiers are ABSENT from file`
+>        - {% icon param-files %} *"Filter out identifiers absent from"*: Select both datasets outputs of **Search in textfiles (grep)** {% icon tool %}.
+>
+> 5. Rename each of the 4 new collections with the four categories: `paired`, `single`, `treat`, `untreat`.
```
Comment (gallardoalba) [inline topics/transcriptomics/tutorials/ref-based/tutorial.md:1358]: It is a bit confusing. As I suggested to you, using inheritable tags (like #paired, #single, #treated, #untreat) could make the tutorial easier to read.

File: topics/transcriptomics/tutorials/ref-based/tutorial.md (hunk 6 above)
Comment (gallardoalba) [reply topics/transcriptomics/tutorials/ref-based/tutorial.md:1411]: Thanks!

File: topics/contributing/tutorials/create-new-tutorial-content/tutorial.md (hunk 9 above)
Comment (lldelisle) [reply topics/contributing/tutorials/create-new-tutorial-content/tutorial.md:968]: I put it into #3389 but not sure it was the best way to do it.

File: news/_posts/2022-04-12-cyot.md (hunk 10 above)
Comment (lldelisle) [reply news/_posts/2022-04-12-cyot.md:2]: I put it into #3389 but not sure it was the best way to do it.

Comment (hexylena) [reply news/_posts/2022-04-12-cyot.md:2]: It's fine for me either way. It was a feature specifically implemented just for this PR, so, it feels less important to split it out. But since @lldelisle opened the PR we can merge that as well and rebase this to remove those commits (but it feels like a lot of work.)

File: topics/transcriptomics/tutorials/ref-based/tutorial.md (hunk 14 above)
Comment (lldelisle) [reply topics/transcriptomics/tutorials/ref-based/tutorial.md:1358]: I agree this is a bit heavy...

File: news/_posts/2022-04-12-cyot.md (hunk 10 above)
Comment (gallardoalba) [reply news/_posts/2022-04-12-cyot.md:2]: If it is related and @hexylena agrees, then it is fine.

File: topics/transcriptomics/tutorials/ref-based/tutorial.md (hunk 14 above)
Comment (lldelisle) [reply topics/transcriptomics/tutorials/ref-based/tutorial.md]: I changed it to tags. @gallardoalba would you mind to review the new paragraph?

Comment (gallardoalba) [reply topics/transcriptomics/tutorials/ref-based/tutorial.md]: Looks great!

**Review by bebatut — COMMENTED** [review]
One last comment. Could you add the level to the metadata? Thanks a lot

**Comment by gallardoalba** [comment]
> > One last comment. Could you add the level to the metadata? Thanks a lot
> 
> Which one would you put? Introductory?

I would say intermediate.

**Review by gallardoalba — DISMISSED** [review]
Great, thanks!

**Comment by bebatut** [comment]
I would have said Introductory for the level. We usually use it at the introduction to RNA-Seq analysis

**Review by bebatut — DISMISSED** [review]
Thanks a lot @lldelisle for all this work!!!! I can not wait to try teaching it

**Comment by lldelisle** [comment]
I still need to update the workflow for DEG but I am a bit fighting...

**Comment by lldelisle** [comment]
Also what should I do with the tour?
How can I test it?

**Review by gallardoalba — DISMISSED** [review]

**Comment by gallardoalba** [comment]
So I guess it can be merged, isn't it?

**Comment by lldelisle** [comment]
Please wait that I update the workflow with tags.

**Comment by gallardoalba** [comment]
> Please wait that I update the workflow with tags.

Ah, ok sorry!

**Comment by lldelisle** [comment]
What do we do with the tour? I update it?

**Comment by lldelisle** [comment]
@shiltemann @hexylena is it possible to hide the tour so I have time to update it after my course?

**Comment by bebatut** [comment]
I think no worries about the tour. I am quite sure it is a lot out of date. We should probably remove it

**Comment by lldelisle** [comment]
Then let's merge like this? Or I remove the tour?

**Comment by lldelisle** [comment]
Hi,
My last changes are:
- Add a sentence about the discrepancies between the results of the first part and the count files on zenodo (this may disappear if we update zenodo).
- Give some hints on how to run the single-read datasets. In my opinion, the single reads datasets should be run with the option in STAR `Length of the genomic sequence around annotated junctions` to 74 as we explain above that it should be read length - 1. However, the counts in zenodo corresponds to STAR ran with 36 like in PE datasets. What do you think?
Thanks

**Comment by lldelisle** [comment]
Hi there,
I am back from holidays and our class is on Wednesday.
1. Should I remove the tour or I keep like this?
2. Should I remove the DEXseq part which is commented but out of date?
Thanks

**Comment by hexylena** [comment]
Let's just remove the tour.  I think ok to leave commented out portions.

**Comment by lldelisle** [comment]
Thanks @hexylena 
Ready for review / merge

**Review by gallardoalba — APPROVED** [review]

**Comment by bebatut** [comment]
The updated tutorial is online :tada: 
I just noticed sth. If nothing is selected for CYOA, then both content is displayed

**Comment by lldelisle** [comment]
@hexylena is there a way to select by default one of them?

**Comment by hexylena** [comment]
Yes, that is expected. I'm not sure there's a good solution for that, picking one by default seems sub-optimal, and having both displayed at least the user can see both simultaneously which feels useful? (Espeially when it's e.g. two separate hands-on boxes). Less so when it's really inline stuff.

**Comment by hexylena** [comment]
We could write some JS to select one (randomly or set a default.)

**Comment by lldelisle** [comment]
I think we can set a default (at least for this tutorial).

**Comment by hexylena** [comment]
What should the default be?

**Comment by hrhotz** [comment]
@bebatut @gallardoalba @hexylena @nsoranzo @jennaj  Thank You very much for all your work and help getting @lldelisle  PR into the GTN

**Comment by lldelisle** [comment]
Yes, thank you very much, I am really happy that we have an up to date tutorial.

**Comment by lldelisle** [comment]
Let see if I have time to do this.
